Update design for "how government works"

jon-kirwan opened the pull request:
## What
https://trello.com/c/vgUN7cPF/631-update-how-government-works-template

- Page needs to be updated to use modern layout classes

## Why

- Consistent look with other pages
- Use design system styles and component guide components
- Fix document outline issues

## Visual changes
<table role="table">
<thead>
<tr>
<th>Before</th>
<th>After</th>
</tr>
</thead>
<tbody>
<tr>
<td valign="top"><a target="_blank" rel="noopener noreferrer" href="https://user-images.githubusercontent.com/87758239/133452337-23a7a53e-6da7-491f-8342-b489edae19ab.png"><img src="https://user-images.githubusercontent.com/87758239/133452337-23a7a53e-6da7-491f-8342-b489edae19ab.png" width="360" style="max-width: 100%;"></a></td>
<td valign="top"><a target="_blank" rel="noopener noreferrer" href="https://user-images.githubusercontent.com/87758239/134308839-a6bae4f5-a50b-4790-87ec-b90c2a5cab45.png"><img src="https://user-images.githubusercontent.com/87758239/134308839-a6bae4f5-a50b-4790-87ec-b90c2a5cab45.png" width="360" style="max-width: 100%;"></a></td>
</tr>
</tbody>
</table>

chris-gds commented on app/views/home/how_government_works.html.erb:
```diff
@@ -1,534 +1,489 @@
+<div class="govuk-width-container">
+  <div class="govuk-grid-row">
+    <div class="govuk-grid-column-two-thirds">
+      <header>
```
There is probably debate to be had about the pros / cons of multiple  `<header>`s on the page - however, following convention of other pages + as there's the `id="global-header"` on the page - I don't _think_ this is needed.

chris-gds commented on app/views/home/how_government_works.html.erb:
```diff
@@ -1,534 +1,489 @@
       </div>
 
-      <div class="full-width-section ministers">
+      <section>
```
I'm glad you used this. I'm not sure we have consistent, recommended agreement on `<section>` use. 

The [Design System layout](https://design-system.service.gov.uk/styles/layout/) doesn't mention it however it is [used here](https://github.com/alphagov/govuk-frontend/blob/4b3c6563a4cffa5b50763231b007b30b60623ea1/app/views/full-page-examples/campaign-page/index.njk) but in a different way.  I've raised a "?" about this [previously](https://github.com/alphagov/gds-way/issues/617) as I'm a bit unclear.

My stance at the moment is that we _should_ be using it [as it benefits some users](https://tink.uk/screen-reader-support-for-html5-sections/) but _how_ we use it I think we need to document a little.  

I would imagine it's OK to use in layout like this but the use above is slightly different and I'm not 100% it's supported.  Trying to find out.

   ```
   <section class="govuk-grid-row">
      <div class="govuk-grid-column-two-thirds">
        <h1 class="govuk-heading-xl">Two-thirds column</h1>
      </div>
    </section>

jon-kirwan replied:
Thanks, will have a look around to see if I can find anything helpful to add to the conversation too

maxgds replied:
I'd tend towards divs until we have a proper strategy around when to use section, but it's probably not worth changing at this point unless it seems to cause any issues.

chris-gds commented on app/views/home/how_government_works.html.erb:
```diff
@@ -1,534 +1,489 @@
         <%= render "govuk_publishing_components/components/title", {
           title: "How government works",
         } %>
         <p class="govuk-body-l">In the UK, the Prime Minister leads the government with the support of the Cabinet and ministers. You can find out <%= link_to("who runs government", "#who-runs-government", class: "govuk-link") %> and <%= link_to("how government is run", "#how-government-is-run", class: "govuk-link") %>, as well as learning about the <%= link_to("history of government", "#history- ...
```
This could be swapped for a [leading paragraph Component](https://components.publishing.service.gov.uk/component-guide/lead_paragraph)

jon-kirwan replied:
Looks like you can use a `sanitize` function with escaped quotes to include html in the text property.

chris-gds commented on app/views/home/how_government_works.html.erb:
```diff
@@ -1,534 +1,489 @@
+         <ul class="govuk-list">
+          <li class="feature-value">
+            <a href="<%= organisations_path(anchor: 'departments') %>" class="feature-value__link">
+              <span class="feature-value__count"><%= @ministerial_department_count %></span>
```
Ah this can be swapped for the newly created ["big numbers"](https://components.publishing.service.gov.uk/component-guide/big_number)

maxgds replied:
Be careful doing this - we previously had to implement code to prevent it overlapping other stuff when users either zoomed in or increased their font size (I don't remember the exact scenario). Test thoroughly for this.

chris-gds replied:
*nods* - [PR Ref FYI ](https://github.com/alphagov/whitehall/pull/5900)

jon-kirwan replied:
Provided empty) alt text for the following images -
```
<%= image_tag @prime_minister.image_url(:s465), { alt: '' } %>
<%= image_tag 'how-gov-works/10_Downing_Street.jpg', alt: '' %>
<%= image_tag 'how-gov-works/Cabinet_01.jpg', alt: '' %>
<%= image_tag 'how-gov-works/Churchill_01.jpg', alt: '' %>
```

jon-kirwan replied:
Okay i've now updated the page to use the big number component. This was a little tricky due to the font size of the number in the component which has meant design changes following a review with Nik.

Zoomed in this looks fine and I can't see any overflow problems on this page. 

Attached a couple of screenshots.

_Infographic 1_

<img src="https://user-images.githubusercontent.com/87758239/134204273-b9c838be-5f06-43af-a2c4-9a6ecf35f88d.png" width="600">

_Infographic 2_

<img src="https://user-images.githubusercontent.com/87758239/134204234-e7f32dfc-c9df-45e4-ab7e-5883713523f9.png" width="600">

chris-gds commented on app/views/home/how_government_works.html.erb:
```diff
@@ -1,534 +1,489 @@
-              heading_level: 3,
-              margin_bottom: 5,
-            } %>
+        <ul class="govuk-list govuk-list--bullet govuk-list--spaced">
```
This looks like it could be swapped to use [List Component ](https://components.publishing.service.gov.uk/component-guide/list)

chris-gds commented on app/views/home/how_government_works.html.erb:
```diff
@@ -1,534 +1,489 @@
-              heading_level: 3,
-              margin_bottom: 5,
-            } %>
+        <ul class="govuk-list govuk-list--bullet govuk-list--spaced">
```
Same here with the [list](https://components.publishing.service.gov.uk/component-guide/list)

chris-gds commented on app/views/home/how_government_works.html.erb:
```diff
@@ -1,534 +1,489 @@
+        <p class="govuk-body">Read about past Prime Ministers, Chancellors and Foreign Secretaries in <%= link_to "notable people", "/government/history#notable-people", class: "govuk-link" %>. Learn more about historic <%= link_to "government buildings", "/government/history#government-buildings", class: "govuk-link" %> on Whitehall and around the UK.</p>
+        <p class="govuk-body govuk-!-margin-bottom-8">You can also find links to <%= link_to "historical research", "/government/history#historical-research", class: "govuk-link" %>, documents and records.</p>
+        
+        <%= image_tag 'how-gov-works/Churchill_01.jpg', alt: 'History of government' %>
```
In the context of this page, I'm not sure about image use in general. However, keeping this in scope of this PR - I think the alt of this needs to change to convey the meaning of this image.  I would go with something like "Winston Churchill" or even "Sir Winston Leonard Spencer Churchill"

maxgds replied:
I'd say in this case it's entirely decorative and not conveying any information whatsoever, so set the alt text to an empty string.

chris-gds replied:
TL;DR - on review I agree. Let's remove alt for all these images on this page.

jon-kirwan replied:
Provided empty) alt text for the following images -
```
<%= image_tag @prime_minister.image_url(:s465), { alt: '' } %>
<%= image_tag 'how-gov-works/10_Downing_Street.jpg', alt: '' %>
<%= image_tag 'how-gov-works/Cabinet_01.jpg', alt: '' %>
<%= image_tag 'how-gov-works/Churchill_01.jpg', alt: '' %>
```

chris-gds commented on app/views/home/how_government_works.html.erb:
```diff
@@ -1,534 +1,489 @@
-        <div class="one-half">
-          <div class="inner-block">
-            <p class="govuk-body">Britain has one of the oldest governments in the world. Find out more about how it has worked and who has shaped it in the <%= link_to "history", "/government/history/", class: "govuk-link" %> section.</p>
+        <ul class="govuk-list govuk-list--bullet govuk-list--spaced">
```
Another list

chris-gds commented on app/views/home/how_government_works.html.erb:
```diff
@@ -1,534 +1,489 @@
+          <p class="govuk-body">The Cabinet is made up of the senior members of government. Every week during Parliament, members of the Cabinet (Secretaries of State from all departments and some other ministers) meet to discuss the most important issues for the government.</p>
+          <p class="govuk-body govuk-!-margin-bottom-8"><%= link_to("See who is in the Cabinet", "/government/ministers", class: "govuk-link") %></p>
+          
+          <%= image_tag 'how-gov-works/Cabinet_01.jpg', alt: 'The Cabinet' %>
```
I don't think this Content is apt, something like "The Cabinet Room at No. 10 Downing Street Cabinet table" I think is more descriptive of what this image is.

jon-kirwan replied:
Provided empty) alt text for the following images -
```
<%= image_tag @prime_minister.image_url(:s465), { alt: '' } %>
<%= image_tag 'how-gov-works/10_Downing_Street.jpg', alt: '' %>
<%= image_tag 'how-gov-works/Cabinet_01.jpg', alt: '' %>
<%= image_tag 'how-gov-works/Churchill_01.jpg', alt: '' %>
```

maxgds commented on app/views/home/how_government_works.html.erb:
```diff
@@ -1,534 +1,553 @@
+          <% if @prime_minister.present? && @prime_minister.image_url(:s465).present? %>
+            <%= image_tag @prime_minister.image_url(:s465), { alt: '', class: 'govuk-!-width-full', width: '465', height: '310' } %>
+          <% else %>
+            <%= image_tag 'how-gov-works/10_Downing_Street.jpg', alt: '', class: 'govuk-!-width-full', width: '465', height: '310' %>
```
These dimensions need attention - I've pushed your branch onto integration to preview it and the images are being stretched too wide on desktop and pushed too narrow on mobile. They seem ok at tablet size. I'm not sure what the interaction causing that is but it definitely needs a fix before we can merge.

jon-kirwan replied:
Yep I see the issue - I've made a few changes to remove the `govuk-!-width-full` class and have replaced with some fluid image styles -
```
img {
  max-width: 100%;
  height: auto;
}
```
which combined with the dimensions seem to work okay now. i've updated the screenshot in the description and also attached a mobile one here

<img src="https://user-images.githubusercontent.com/87758239/134308604-9a2d34b3-be49-4ac2-894e-f0b1d338e2ad.png" width="300">

maxgds commented on test/functional/home_controller_test.rb:
```diff
@@ -18,14 +18,14 @@ class HomeControllerTest < ActionController::TestCase
     get :how_government_works
 
-    assert_select ".prime-minister p a", "Firstname Lastname"
+    assert_select ".prime-minister > p:nth-child(5) [href]", "Firstname Lastname"
```
Not worth changing for this PR but a thought for future: this is not a particularly robust test. If we start adding extra children or removing some of the current content then that target of 5th child becomes invalid and the test needs an update. Probably better to tie things together more with classes generally, I'd have gone for BEM classes across the prime minister metadata myself. At that point you can both style and test the content more directly.

jon-kirwan replied:
Thanks Max - good point. I'll bear that in mind for future.

maxgds commented:
I think this is looking ok now - if you can kill the temporary commit and rebase against master I'll re-review.

jon-kirwan commented:
> I think this is looking ok now - if you can kill the temporary commit and rebase against master I'll re-review.

Sorry one final unexpected thing following the rebase is that one of the tests failed.
```
expected not to find visible css ".feature-ministers", found 1 match: "0   Ministerial departments\n0   Non-ministerial departments\n300+   Agencies other public bodies" (RSpec::Expectations::ExpectationNotMetError)
./features/step_definitions/sitewide_settings_steps.rb:24:in `/^I should not see the minister counts$/'
./features/support/admin_taxonomy_helper.rb:31:in `block in <top (required)>'
features/sitewide_settings.feature:6:in `I should not see the minister counts'
```

This is because I deleted the `.featured-ministers` selector (it's no longer used for any style purpose but is used by the test). I've now added this back in to the reshuffle / ministers section which has fixed the issue... but just wanted to mention it.
